code/espurna/light.cpp

```diff
@@ -607,27 +647,106 @@ unsigned int _toPWM(unsigned char id) {
     return _toPWM(_light_channels[id].current, useGamma, _light_channels[id].inverse);
 }
 
-void _lightTransition(unsigned long step) {
+namespace {
 
-    // Transitions based on current step. If step == 0, then it is the last transition
-    for (auto& channel : _light_channels) {
-        if (!step) {
-            channel.current = channel.target;
-        } else {
-            channel.current += (double) (channel.target - channel.current) / (step + 1);
+class ChannelTransition {
+public:
+    struct Transition {
+        float& value;
+        unsigned char target;
+        float step;
+        size_t count;
+
+        void debug() const {
+            DEBUG_MSG_P(PSTR("[LIGHT] Transition from=%s to=%u step=%s times=%u\n"),
+                String(value).c_str(), target, String(step).c_str(), count);
         }
+    };
+
+    explicit ChannelTransition(std::vector<channel_t>& channels, unsigned long time, unsigned long step) :
+        _time(time),
+        _step(step)
+    {
+        bool action { false };
+        for (auto& channel : channels) {
+            float diff = static_cast<float>(channel.target) - channel.current;
+            if (!time || (std::abs(diff) <= std::numeric_limits<float>::epsilon())) {
+                channel.current = channel.target;
+                continue;
+            }
```

Comment:
Also reworked 'empty' condition. Either time is 0 (transitions disabled) or transition is too small - it simply does nothing, but still triggers the channel update